Optimized DeepSeek-v2 on Gaudi

gyou2021 opened the pull request:
1. Optimized MLA attention for inference;
2. Optimzed MoE with dynamic MoE and expert slice for inference;
3. Optimzed cache for inference;
4. Optimized training.

Inference example:
python run_generation.py \
        --model_name_or_path <DeepSeek-V2-Lite directory> \
        --bf16 \
        --trim_logits \
        --batch_size 2 \
        --max_input_tokens 128 \
        --max_new_tokens 128 \
        --use_hpu_graphs \
        --use_kv_cache \
        --reuse_cache \
        --bucket_size 128 \
        --bucket_internal \
        --limit_hpu_graphs \
        --use_flash_attention

Training example:
python3 ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py
--config_name deepseek_v2(model_name_or_path for your local)
--tokenizer_name deepseek_v2(model_name_or_path for your local)
--dataset_name alpaca  
--block_size 4096
--do_train
--num_train_epochs 1
--max_steps 10
--per_device_train_batch_size 1
--gradient_accumulation_steps 1
--use_flash_attention True
--attn_softmax_bf16 False
--gradient_checkpointing
--learning_rate 2.4e-4
--gaudi_config_name gaudi_config.json(just like Habana/gpt2)
--bf16
--save_strategy no
--no_save_last_ckpt
--output_dir path_to_out
--overwrite_output_dir
--report_to tensorboard
--logging_strategy steps
--logging_dir path_to_log
--logging_steps 1
--evaluation_strategy no
--use_habana
--use_lazy_mode
--deepspeed zero.json
 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

emascarenhas commented:
@gyou2021 ,
There are errors in style. Please run "make style" and fix the errors that are reported.

Do you need to add or modify tests?

Suggest running all the text generation slow tests with and without change so that no new issues are seen from this change. Please post results here.
GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py

Is there any impact on fine tuning or training scripts?
Are any changes to README required per the command line?

Edward

xuguangxin commented:
@mengker33 @xhaihao @hlin99 please help review too. 
Thank you.

ranzhejiang commented:
> @gyou2021 , There are errors in style. Please run "make style" and fix the errors that are reported.
> 
> Do you need to add or modify tests?
> 
> Suggest running all the text generation slow tests with and without change so that no new issues are seen from this change. Please post results here. GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py
> 
> Is there any impact on fine tuning or training scripts? Are any changes to README required per the command line?
> 
> Edward

Thank you for review, for training scripts, we just only add some missing fused kernel config  arguments for run_clm.py following the already existing run_lora_clm.py.

gyou2021 commented:
> @gyou2021 , There are errors in style. Please run "make style" and fix the errors that are reported.
> 
> Do you need to add or modify tests?
> 
> Suggest running all the text generation slow tests with and without change so that no new issues are seen from this change. Please post results here. GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py
> 
> Is there any impact on fine tuning or training scripts? Are any changes to README required per the command line?
> 
> Edward
Thank you for your comments.
Fixed style errors .  
No, it is not needed to add or modify tests for deepseek-v2 inference since the PR is  about its optimization.   
Updated DeepSeek-v2 in README.
The Deepseek-V2-Lite test case passed the pytest.

gyou2021 commented:
> > @gyou2021 , There are errors in style. Please run "make style" and fix the errors that are reported.
> > Do you need to add or modify tests?
> > Suggest running all the text generation slow tests with and without change so that no new issues are seen from this change. Please post results here. GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py
> > Is there any impact on fine tuning or training scripts? Are any changes to README required per the command line?
> > Edward
> 
> Thank you for review, for training scripts, we just only add some missing fused kernel config arguments for run_clm.py following the already existing run_lora_clm.py.

Thank you for your comments.
Fixed style errors .
No, it is not needed to add or modify tests for deepseek-v2 inference since the PR is about its optimization.
Updated DeepSeek-v2 in README.
The Deepseek-V2-Lite test case passed the pytest.

emascarenhas commented on examples/language-modeling/run_clm.py:
```diff
@@ -156,6 +156,40 @@ class ModelArguments:
+        default=False,
+        metadata={
+            "help": (
+                "Whether to run attention softmax layer in bf16 precision for fine-tuning. The current support is limited to Llama only."
```
What does it mean to make the statement here about Llama when the change is for the DeepSeek-v2 model?

ranzhejiang replied:
> What does it mean to make the statement here about Llama when the change is for the DeepSeek-v2 model?

In fact, I just copied these statements from run_lora_clm.py, which introduced some outdated code. Our customer needs this training script as an example in their workload, so we need to add more statements for some missing arguments. Now I have removed these outdated statements. BTW, these parameters are not only used for LLaMA, but can also be applied to Qwen, Mixtral, ChatGLM and other similar models.

emascarenhas commented on examples/language-modeling/run_clm.py:
```diff
@@ -156,6 +156,40 @@ class ModelArguments:
+        metadata={
+            "help": (
+                "Whether to use Habana flash attention for fine-tuning. The current support is limited to Llama only."
+            )
```
Same question as line 163.

emascarenhas commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -20,37 +20,80 @@
+
+try:
+    from habana_frameworks.torch.hpex.kernels import FusedSDPA
+except ImportError:
```
Do you need to print that FusedSDPA will be used here?

gyou2021 replied:
Yes. When FusedSDPA is not available, performance may decrease. The printed message can remind the user to install the package to avoid it.

emascarenhas commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -649,107 +961,331 @@ def _init_rope(self):
-        deltas are:
-        - add  token_idx
-        - optimize KV cache
+        Attention masks and past cache are removed.
```
Any reason to remove the Copied from comment here?

gyou2021 replied:
The attention algorithm is changed to an optimized MLA, not KV cache optimization.

emascarenhas commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -797,13 +1342,18 @@ def forward(
-        The deltas are:
-        - add token_idx
-        """
         """
```
Copied from comment removed?

gyou2021 replied:
Due to changes in the attention algorithm, its API has been updated, affecting how DeepSeekV2DecoderLayer calls it. This is not related to KV cache optimization, so the comment was removed.

gyou2021 commented:
> Hi @gyou2021 , Can you provide a command line you expect to work with Deepseek-v2 and with run_clm.py?

python3 ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py
--config_name deepseek_v2
--tokenizer_name deepseek_v2
--dataset_name alpaca  
--block_size 4096
--do_train
--num_train_epochs 1
--max_steps 10
--per_device_train_batch_size 1
--gradient_accumulation_steps 1
--use_flash_attention True
--attn_softmax_bf16 False
--learning_rate 2.4e-4
--gaudi_config_name gaudi_config.json
--bf16
--save_strategy no
--no_save_last_ckpt
--output_dir path_to_out
--overwrite_output_dir
--report_to tensorboard
--logging_strategy steps
--logging_dir path_to_log
--logging_steps 1
--evaluation_strategy no
--use_habana
--use_lazy_mode

emascarenhas commented:
I could not run the command provided as is since the config file, etc weren't local. Here is a command that works which will pull the model from HF. Similar commands using other Huggingface DeepSeek models should also work.

python3 ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py \
--config_name deepseek-ai/DeepSeek-V2-Lite \
--tokenizer_name deepseek-ai/DeepSeek-V2-Lite \
--dataset_name tatsu-lab/alpaca \
--block_size 4096 \
--do_train \
--num_train_epochs 1 \
--max_steps 10 \
--per_device_train_batch_size 1 \
--gradient_accumulation_steps 1 \
--use_flash_attention True \
--attn_softmax_bf16 False \
--learning_rate 2.4e-4 \
--gaudi_config_name Habana/gpt2 \
--bf16 \
--save_strategy no \
--no_save_last_ckpt \
--output_dir path_to_out \
--overwrite_output_dir \
--report_to tensorboard \
--logging_strategy steps \
--logging_dir path_to_log \
--logging_steps 1 \
--evaluation_strategy no \
--use_habana \
--use_lazy_mode

ranzhejiang commented:
> I could not run the command provided as is since the config file, etc weren't local. Here is a command that works which will pull the model from HF. Similar commands using other Huggingface DeepSeek models should also work.
> 
> python3 ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py --config_name deepseek-ai/DeepSeek-V2-Lite --tokenizer_name deepseek-ai/DeepSeek-V2-Lite --dataset_name tatsu-lab/alpaca --block_size 4096 --do_train --num_train_epochs 1 --max_steps 10 --per_device_train_batch_size 1 --gradient_accumulation_steps 1 --use_flash_attention True --attn_softmax_bf16 False --learning_rate 2.4e-4 --gaudi_config_name Habana/gpt2 --bf16 --save_strategy no --no_save_last_ckpt --output_dir path_to_out --overwrite_output_dir --report_to tensorboard --logging_strategy steps --logging_dir path_to_log --logging_steps 1 --evaluation_strategy no --use_habana --use_lazy_mode

Thank you for testing, and now it can work with your command, if you meet  OOM, you can add --deepspeed zero3.json  or  --gradient_checkpointing  to reduce memory.

mengker33 commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -835,14 +1385,27 @@ def forward(
         residual = hidden_states
         hidden_states = self.post_attention_layernorm(hidden_states)
-        hidden_states = self.mlp(hidden_states)
+        if isinstance(self.mlp, DeepseekV2MoE):
```
self.mlp is already initialized with different class according to the config, so we don't need this here, right?

gyou2021 replied:
Yes.

gyou2021 commented:
We added the training command you verified into the README.txt of language modeling.


> I was able to run training successfully with the additional options suggested. The command appears to hang if the deepspeed config is not passed, it should have errored out with OOM. Here is a command that completes successfully. python3 ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py --config_name deepseek-ai/DeepSeek-V2-Lite --tokenizer_name deepseek-ai/DeepSeek-V2-Lite --dataset_name tatsu-lab/alpaca --block_size 4096 --do_train --num_train_epochs 1 --max_steps 10 --per_device_train_batch_size 1 --gradient_accumulation_steps 1 --use_flash_attention True --attn_softmax_bf16 False --gradient_checkpointing --learning_rate 2.4e-4 --gaudi_config_name Habana/gpt2 --bf16 --save_strategy no --no_save_last_ckpt --output_dir /root/deepseek-v2-lite --overwrite_output_dir --logging_strategy steps --logging_dir /root/deepseek-v2-lite/log --logging_steps 1 --evaluation_strategy no --use_habana --use_lazy_mode --deepspeed llama2_ds_zero3_config.json

regisss commented on README.md:
```diff
@@ -256,7 +256,7 @@ The following model architectures, tasks and device distributions have been vali
 | MiniCPM3 |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Baichuan2 | <div style="text-align:left"><li>DeepSpeed</li></div> | <div style="text-align:left"><li>Single card</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
-| DeepSeek-V2 |   | :heavy_check_mark: | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
+| DeepSeek-V2 | :heavy_check_mark: | :heavy_check_mark: | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
Can you add the same to https://github.com/huggingface/optimum-habana/blob/3c251d0480d916bcb08d0bda092edf03847ca11b/docs/source/index.mdx?plain=1#L59 please?

gyou2021 replied:
Added. Thank you.

regisss commented on examples/language-modeling/README.md:
````diff
@@ -184,6 +184,36 @@ python ../gaudi_spawn.py \
     --logging_steps 20
 ```
 
+### Multi-card Training with Deepspeed (DeepSeek-V2-Lite)
+```bash
+python ../gaudi_spawn.py --world_size 8 --use_deepspeed run_clm.py
+    --config_name deepseek-ai/DeepSeek-V2-Lite
+    --tokenizer_name deepseek-ai/DeepSeek-V2-Lite
+    --dataset_name tatsu-lab/alpaca
+    --block_size 4096
+    --do_train
+    --num_train_epochs 1
+    --max_steps 10
+    --per_device_train_batch_size 1
+    --gradient_accumulation_steps 1
+    --use_flash_attention True
+    --attn_softmax_bf16 False
+    --gradient_checkpointing
+    --learning_rate 2.4e-4
+    --gaudi_config_name Habana/gpt2
+    --bf16
+    --save_strategy no
+    --no_save_last_ckpt
+    --output_dir /root/deepseek-v2-lite
+    --overwrite_output_dir
+    --logging_strategy steps
+    --logging_dir /root/deepseek-v2-lite/log
+    --logging_steps 1
+    --evaluation_strategy no
+    --use_habana
+    --use_lazy_mode
+    --deepspeed llama2_ds_zero3_config.json
+```
````
IS there any specific arg here that is specific to this model? Otherwise let's just try to keep the README concise.

gyou2021 replied:
Certain arguments, such as --use_flash_attention, enable optimization during training. Some of these argument values were derived from our experiments. The command has been verified on both the local server and DevCloud. If any arguments are missing, the training process may fail. To ensure successful execution, we have included these details in the README.

regisss commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -20,37 +20,80 @@
+# This makes `_prepare_4d_causal_attention_mask` a leaf function in the FX graph.
+# It means that the function will not be traced through and simply appear as a node in the graph.
+if is_torch_fx_available():
+    if not is_torch_greater_or_equal_than_1_13:
+        import torch.fx
```
We don't need that since SynapseAI relies on Torch > 1.13

gyou2021 replied:
Removed. Thank you.

regisss commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -68,6 +111,89 @@ def _get_unpad_data(attention_mask):
+    attention_mask: Optional[torch.Tensor] = None,
+) -> Union[torch.Tensor, int]:
+    r"""
+    Computes auxiliary load balancing loss as in Switch Transformer - implemented in Pytorch.
+
+    See Switch Transformer (https://arxiv.org/abs/2101.03961) for more details. This function implements the loss
+    function presented in equations (4) - (6) of the paper. It aims at penalizing cases where the routing between
+    experts is too unbalanced.
+
+    Args:
+        gate_logits:
+            Logits from the `gate`, should be a tuple of model.config.num_hidden_layers tensors of
+            shape [batch_size X sequence_length, num_experts].
+        num_experts:
+            Number of experts
+        top_k:
+            The number of experts to route per-token, can be also interpreted as the `top-k` routing
+            parameter.
+        attention_mask (`torch.Tensor`, *optional*):
+            The attention_mask used in forward function
+            shape [batch_size X sequence_length] if not None.
+
+    Returns:
+        The auxiliary loss.
+    """
```
Can you explain how this method differs from the original?

gyou2021 replied:
The auxiliary load balancing loss here is a mechanism designed to ensure that the computational load is evenly distributed among the available experts. This loss function penalizes scenarios where the routing of tokens to experts is highly imbalanced, encouraging a more uniform distribution of tokens across the experts.
The goal is to prevent any single expert from being overloaded while others remain underutilized. This helps in maintaining efficiency and stability during training and inference.

regisss commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -476,67 +626,209 @@ def __init__(self, config):
         if config.n_shared_experts is not None:
             intermediate_size = config.moe_intermediate_size * config.n_shared_experts
             self.shared_experts = DeepseekV2MLP(config=config, intermediate_size=intermediate_size)
+        SLICE_MAX_EXPERT = 80
```
Should be set at the top of the file. Why is it hardcoded?

regisss reviewed (commented):
I left a few comments.
Also, have you checked that [this test](https://github.com/huggingface/optimum-habana/blob/3c251d0480d916bcb08d0bda092edf03847ca11b/tests/test_text_generation_example.py#L58) still passes?

gyou2021 commented:
> [this test](https://github.com/huggingface/optimum-habana/blob/3c251d0480d916bcb08d0bda092edf03847ca11b/tests/test_text_generation_example.py#L58) still passes?

Yes. Our tests verified [this test](https://github.com/huggingface/optimum-habana/blob/3c251d0480d916bcb08d0bda092edf03847ca11b/tests/test_text_generation_example.py#L58) about DeepSeek-V2-Lite still passes.

regisss commented on optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py:
```diff
@@ -62,16 +62,22 @@
+#         import torch.fx
 
-    _prepare_4d_causal_attention_mask = torch.fx.wrap(_prepare_4d_causal_attention_mask)
+#    _prepare_4d_causal_attention_mask = torch.fx.wrap(_prepare_4d_causal_attention_mask)
```
Let's remove the commented lines

regisss reviewed (commented):
You'll have to run `make syle` to make the style check pass

regisss commented:
@gyou2021 There is also a merge conflict to solve

gyou2021 commented:
> You'll have to run `make syle` to make the style check pass

It has been run and style problems were fixed.

gyou2021 commented:
> @gyou2021 There is also a merge conflict to solve

Solved. Thank you.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1677). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
